// time-pray/dist/src/method-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodsData = exports.defaultParams = exports.asrFactors = void 0;
exports.asrFactors = { Standard: { factor: 1 }, Hanafi: { factor: 2 } };
const setting = {
    imsak: { minutes: 10 },
    dhuhr: { minutes: 0 },
    asr: exports.asrFactors.Standard,
    highLats: "NightMiddle",
};
exports.defaultParams = {
    maghrib: { minutes: 0 },
    midnight: "Standard",
    ...setting,
};
exports.MethodsData = {
    MWL: {
        name: "Muslim World League",
        params: {
            ...exports.defaultParams,
            fajr: { degree: 18 },
            isha: { degree: 17 },
        },
    },
    ISNA: {
        name: "Islamic Society of North America (ISNA)",
        params: {
            ...exports.defaultParams,
            fajr: { degree: 15 },
            isha: { degree: 15 },
        },
    },
    Egypt: {
        name: "Egyptian General Authority of Survey",
        params: {
            ...exports.defaultParams,
            fajr: { degree: 19.5 },
            isha: { degree: 17.5 },
        },
    },
    Makkah: {
        name: "Umm Al-Qura University, Makkah",
        params: {
            ...exports.defaultParams,
            fajr: { degree: 18.5 },
            isha: { minutes: 90 },
        },
    },
    Karachi: {
        name: "University of Islamic Sciences, Karachi",
        params: {
            ...exports.defaultParams,
            fajr: { degree: 18 },
            isha: { degree: 18 },
        },
    },
    Tehran: {
        name: "Institute of Geophysics, University of Tehran",
        params: {
            ...exports.defaultParams,
            fajr: { degree: 17.7 },
            isha: { degree: 14 },
            maghrib: { degree: 4.5 },
            midnight: "Jafari",
        },
    },
    Jafari: {
        name: "Shia Ithna-Ashari, Leva Institute, Qum",
        params: {
            ...exports.defaultParams,
            fajr: { degree: 16 },
            isha: { degree: 14 },
            maghrib: { degree: 4 },
            midnight: "Jafari",
        },
    },
};
